### PR TITLE
os/bluestore: fix bluefs log run out of space

### PR DESCRIPTION
**Pull request by kitten1007**
Fixes: https://tracker.ceph.com/issues/51217
Signed-off-by: Shu Yu <yushu20171007@163.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by aclamk** [comment]
When we replay BlueFS log, we read a sequence of bluefs_transaction_t elements.
Each transaction must be read completely before we can proceed with interpretation.

BlueFS log is dynamic. It grows when new space is needed.
When writing new transaction to log it is critical that entire transaction fits inside disk space allocated for log.
Obviously, if transaction is larger then space, we assert.

Unfortunately, it is impossible to extent BlueFS log in the same transaction that exceeds previous allocated space.
Imagine following sequence:
1) there is 1000 more bytes allocated for log
2) bluefs_transaction_t to write is 2000 bytes long
3) additional 10000 bytes is allocated
4) transaction is updated to reflect additional space in log 
5) now log has 11000 bytes available
6) log is written

Now we have a situation of #35776 - information of log extension will be available for _replay function, *after* it will need it.

In conclusion:
*** We need to make sure additional BlueFS log extents are added to transaction that is already stable on disk, before we use it. ***


NOTE: Luminous Ceph - v12.2.12 does not seem to have #35473 backported, so it is still possible to generate excessive log size problem https://tracker.ceph.com/issues/46195 .

**Review by aclamk — CHANGES_REQUESTED** [review]
I am against adding logic in this form, as it does not fix underlying problem of extending BlueFS log.
BlueFS log space must be already stable in some previous transaction before we can use it. It is an error to add and use allocated space in the same transaction.

I imagine that some special transaction op "OP_ADD_EXTENTS" can be a solution. Any time we detect that our runway is too small, we will inject small bluefs_transaction_t with only one op that extends bluefs log (ino 1).
